Whitelabelling users font family font setting

adam-james-v opened the pull request:
This PR addresses a change breaking stats. 

#### Related:
- Original PR: #22612
- Reverted by: #22766

## Problem
The original PR hardcoded the font-path as `(def font-path "./resources/frontend_client/app/fonts/")` which is a relative path as a string. In a REPL and in CI which runs source, this path works correctly.

However, once Metabase is built, everything runs from an uberjar and (naturally) such a file path will not work correctly. The resources do correctly get bundled inside the jar, but they were not being accessed. The implementation in fact always looked in the regular filesystem.

## Solution
This PR addresses the issue by using `metabase.util.files/with-open-path-to-resource` which ends up using `clojure.java.io/resource`; the correct way to grab resources.

### Verifying
To verify that this does actually work, you can build and run the jar:

In the project root directory, build an Enterprise uberjar:
`MB_EDITION=ee ./bin/build`

wait for the build to complete, then move the jar outside of the project:
`mv target/uberjar/metabase.jar ~/Downloads` (or some other directory)

Run the jar:
`MB_EDITION=ee java -jar ~/Downloads/metabase.jar`

Then, check that Metabase is working. The old PR would not even load properly as the resources weren't available, so even seeing Metabase load properly in your browser indicates success. But, just to be safe, head to Admin Settings > Appearance (the bottom of the menu) and try changing the font.

adam-james-v committed:
Add font utilities to derive list of available fonts from font dirs

This PR adds two defsettings to facilitate Enterprise users' ability to set the font for their instance. The
'available-fonts' can be grabbed by the frontend to populate the dropdown menu. The 'application-font' setting can be
written by the frontend to save the chosen font.

npfitz committed:
Adding some fonts and font selector

adam-james-v committed:
Send  (("dirname", "font name"),) to frontend

Instead of only sending a list of names, send a tuple of `[directory-name, Display Name]`

npfitz committed:
Adding half of the fonts

adam-james-v committed:
Font list derived from font directory/font file names

This is an attempt to derive the 'Display Name' of a font by comparing the directory name to the file names. '_' is
replaced with spaces, but '-' exists in the 'Lato' font directory name, and we may have naming confusion in the
future. Basically, I want to be liberal with what font directory/file names we can receive.

Not certain this is the right approach yet, but it will work and correctly provides names based on the fonts available.

npfitz committed:
Adding remaining fonts

npfitz committed:
Using Default Font family in visual components

npfitz committed:
Removing Noto Sans JP

adam-james-v committed:
Small change to fix some lint warnings

adam-james-v committed:
Move logging

adam-james-v committed:
Alter hardcoded font path

npfitz committed:
Added cypress test

adam-james-v committed:
Change Cypress Test to look for correct text

npfitz committed:
adding visual test

npfitz committed:
fixing cypress test

adam-james-v committed:
Added unit tests for fonts.clj, and simplified creating display name

The display name approach is simple -> underscores become spaces, and names are split on dashes/first word is taken
from the split.

This might not be robust to future font installations, but will be a good v1

adam-james-v committed:
Actually add the unit tests

adam-james-v committed:
Fix lint problems

adam-james-v committed:
Add setter to application-font to prevent invalid font setting

adam-james-v committed:
Addressing PR review points

- fixed 'normalize-font-dirname' description
- 'contains-font-file?' returns true/false
- 'available-fonts-test' now properly uses 'is'

adam-james-v committed:
Make font-path private

adam-james-v committed:
Lint error.

npfitz committed:
PR Cleanup

adam-james-v committed:
Modified fonts to use java.io/resource to hopefully work in a jar

adam-james-v committed:
Simplified Lato folder name

metamben commented on src/metabase/util/fonts.clj:
```diff
@@ -0,0 +1,38 @@
+(defn- available-fonts*
+  []
+  (u.files/with-open-path-to-resource [font-path "frontend_client/app/fonts"]
+    (let [font-path-str (str (.toString font-path) "/")]
```
Calling `.toString` seems redundant.

metamben commented on src/metabase/util/fonts.clj:
```diff
@@ -0,0 +1,38 @@
+           (filter contains-font-file?)
+           (map #(str/replace (str %) font-path-str ""))
+           (map normalize-font-dirname)
+           (sort-by #(str/lower-case %))))))
```
`#(str/lower-case %)` -> `str/lower-case`

adam-james-v committed:
Address some backend feedback

adam-james-v committed:
Lint error from reflection with .toString on a path

I removed the explicit .toString inside the lamdba functino and str/includes still works.

adam-james-v committed:
no longer need u.files in test ns
